feat(shared-data,api,robot-server): Use new OT-3 deck slot naming style

SyntaxColoring opened the pull request:
# Overview

This is a chunk of back-end support for OT-3 style deck slot names, like `D1` instead of `1`.

Closes RLAB-250, RLAB-251, RLAB-253, RLAB-256, RLAB-257, RLAB-339, and RLAB-346.

# Changelog

In general, the robot will now accept deck slots of *either style* (`1` or `D1`) as input. It will normalize them to the  device's preferred style (`1` for OT-2s, `D1` for OT-3s) in its internal state, and in its output. In protocol analyses, the style in the output will follow the protocol's declared `robotType`.

This affects:

* `loadLabware` commands
* `moveLabware` commands
* `loadModule` commands
* Labware offsets `POST`ed over HTTP
* Run summaries visible over HTTP

The underlying OT-3 deck definition in `shared-data` is changed to use the OT-3-style names. The module definitions are updated to match.

# Review requests

* [ ] Scrutinize my updates in `shared-data`.
  * [ ] Make sure I'm updating things to the *correct* new values. [Here's a handy quick reference table](https://gist.github.com/SyntaxColoring/9139aa7420e3993d20e9028501f65755), but double-check this with a physical OT-3.
  * [ ] Make sure I didn't accidentally leave any of the OT-2-style numbers behind. **It's very important that the module definitions and deck definitions are consistent with each other.**
* [ ] Can you think of any other places that might need to normalize deck slots before comparing them? For example, any conflict-checking code hiding out somewhere?

# Risk assessment

High.

This unfortunately has to change a bunch of stuff all at once, because it's all cross-dependent on each other.

Code that does this is in danger of breaking silently:

```
if some_slot == some_other_slot:
    ...
```

We do this in places such as:

* Determining the offset of a module. Modules have per-slot transformations, which are matched by slot ID.
* Determining the Labware Position Check offset to apply to a labware. Labware Position Check offsets are defined by a labware URI and a location, and the location includes the slot ID.

# Test plan

[Here's the protocol file that I'm using.](https://gist.github.com/SyntaxColoring/f4419bdf85eb73acb951a5ff45e811e8)

After pushing this branch to an OT-3, you'll need to factory-reset runs history before the robot is usable. You'll also need to factory-reset runs history when pushing `edge` back.

* [x] OT-3-only: Make sure you can still detach and attach pipettes.
* [x] OT-3-only: Make sure you can still calibrate pipettes.
* [x] OT-2 and OT-3: Make sure the Labware Position Check flow continues to behave correctly.
    * Yes, with PRs #12822 and #12815.
* [x] OT-2 and OT-3: Make sure that offsets saved in Labware Position Check correctly apply to the actual protocol run.
* [x] OT-2 and OT-3: Make sure you can apply labware offsets from prior runs.
* [x] OT-2 and OT-3: Make sure that in a protocol run, the robot physically continues to go to the right place, especially when the labware is loaded atop a module.
* [x] OT-2 and OT-3: Make sure the robot continues to dodge Thermocyclers, in protocols that have them.
* [x] OT-2 and OT-3: In general, make sure the Opentrons App shows slot names like "1" everywhere it's dealing with an OT-2, and "D1" everywhere it's dealing with an OT-3.
    * RAUT-473 is one missed spot, but otherwise this seems good.
* [x] OT-2 and OT-3: In general, make sure the Opentrons App keeps displaying the deck map correctly.

SyntaxColoring committed:
Update OT-3 deck definition IDs to be OT-3-style.

SyntaxColoring committed:
Update module definitions to refer to new OT-3 deck IDs.

SyntaxColoring committed:
Remove meddlesome assertion.

SyntaxColoring committed:
Hack together some canonicalization for load_labware.

SyntaxColoring committed:
Remove internal stringification from legacy resolve_module_location().

SyntaxColoring committed:
Update command schema.

SyntaxColoring committed:
Fix module offset test.

SyntaxColoring committed:
Canonicalize loadModule commands.

SyntaxColoring committed:
canonicalize->normalize

SyntaxColoring committed:
Add integration test for normalization over HTTP.

SyntaxColoring committed:
Fix bug in loadModule command normalization.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

SyntaxColoring committed:
Fix command turning itself into a non-setup command.

SyntaxColoring committed:
Normalize labware offset requests.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -16,6 +17,17 @@
 LoadLabwareCommandType = Literal["loadLabware"]
 
 
+def canonicalize_location(
```
To do: Move to a free function somewhere under `opentrons.protocol_engine` that we can mock out in `ProtocolEngine`'s tests.

SyntaxColoring replied:
To do: Also normalize `moveLabware` commands.

SyntaxColoring commented on api/src/opentrons/types.py:
```diff
@@ -223,6 +225,19 @@ class DeckSlotName(enum.Enum):
     SLOT_11 = "11"
     FIXED_TRASH = "12"
 
+    SLOT_A1 = "A1"
+    SLOT_A2 = "A2"
+    SLOT_A3 = "A3"
+    SLOT_B1 = "B1"
+    SLOT_B2 = "B2"
+    SLOT_B3 = "B3"
+    SLOT_C1 = "C1"
+    SLOT_C2 = "C2"
+    SLOT_C3 = "C3"
+    SLOT_D1 = "D1"
+    SLOT_D2 = "D2"
+    SLOT_D3 = "D3"
+
```
Making `DeckSlotName` contain both styles like this feels icky, but I think it's our best option, at least for now.

The public models, for JSON protocols and HTTP requests, all use this enum. Those models need to support both styles now. Therefore, our options are either:

1. Make this enum contain both styles, like I've done here.
2. Or, add a second enum, and do `Union[OT2DeckSlotName, OT3DeckSlotName]` everywhere.

The second one seems basically functionally equivalent to the first, except that it's more work and more diff. :)

I think it would be more worthwhile to have two separate unions is if we made a bunch of our conflict-checking logic generic. Like `DeckConflictChecker[OT2DeckSlotName].check(...)`. But that's beyond the scope of this PR, and leaving this as a single enum doesn't make that any harder to do in the future.

jbleon95 replied:
I think for purposes of keeping the footprint of these changes as small as possible, having `DeckSlotName` contain both styles is adequate for the purposes of this feature. The only thing that is in the back of my mind regarding this is if we do any equality checking between the `DeckSlotName` enum and if this would break any of that, but to my knowledge we don't.

TamarZanzouri replied:
Maybe we should just add a TODO to think of an alternative when we get a chance? for the sake of this PR its ok by me to leave this as is but would love to find a better solution.

SyntaxColoring replied:
I discussed this in-person with @jbleon95.

I think we should have 2 or 3 different enums:

* `DeckSlotName`.
    * This is the public type, exposed over HTTP and in the command schema.
    * It has a 1:1 correspondence with `id`s in deck definitions.
    * We can optionally split this into `OT2DeckSlotName` and `OT3DeckSlotName`, if that turns out to be helpful. But I don't think this is the real problem.
* `UnifiedDeckSlotName`.
    * The values are the equivalence classes of deck slots. For example, `UnifiedDeckSlotName.D1_OR_1`.
    * This should be used internally, basically everywhere in Protocol Engine. At Protocol Engine's interface boundaries, Protocol Engine should convert between this and `DeckSlotName`.
    * Perhaps the Python Protocol API internals should also use `UnifiedDeckSlotName`. I'm not sure.

This would help address the hazards that seem to be cropping up in practice:

* Various things are hard-coding slots. For example, Thermocycler-dodging logic has a hard-coded list of slot pairs where dodging needs to happen, and a hard-coded "middle slot" to use when planning the motion of the dodge itself. These things are currently in danger of accidentally assuming a specific deck definition. They should be rewritten to use `UnifiedDeckSlotName` to be definition-agnostic.
* Various things are comparing deck slots for equality. For example, checking for conflicts between `loadLabware`s, and checking which labware offset applies to a `loadLabware`. These things are currently in danger of false-negatives when we accidentally use a mixture of OT-2 and OT-3 slot names inside Protocol Engine state. To avoid this, Protocol Engine state should be rewritten to store `UnifiedDeckSlotName`s instead, so it's impossible to create that mixture.

This looks like it'll be too big of a refactor to fit in this PR. For example, [`LoadedLabware`](https://github.com/Opentrons/opentrons/blob/b2dbd3c2b62ba7771682126f50e4ea7f3de84c2b/api/src/opentrons/protocol_engine/types.py#L488) will need to be split into an HTTP-exposed model (with `DeckSlotName`) and an internal model (with `UnifiedDeckSlotName`).

SyntaxColoring replied:
Discussed the above with @TamarZanzouri:

Typing out `UnifiedSlotName.D1_OR_1` might be onerous and clunky. Maybe instead of doing that, it's just a set with nameless values, and we provide lookup functions like `get_unified_slot(ot3_name="D1")`.

TamarZanzouri replied:
I like this! I feel like its more verbose and maintainable. for sure not for this PR though

SyntaxColoring replied:
https://opentrons.atlassian.net/browse/RSS-259

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Merging to incorporate https://github.com/Opentrons/opentrons/pull/12595.

SyntaxColoring committed:
use words more good

SyntaxColoring committed:
by the robot type -> for the robot type

(shall not perish from the earth)

SyntaxColoring committed:
Reconcile Jeremy's DeckSlotName changes with mine.

SyntaxColoring committed:
Add some more internal DeckSlotName documentation.

SyntaxColoring committed:
Leave todos on Deck methods that look private.

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/command.py:
```diff
@@ -85,6 +93,11 @@ class BaseCommandCreate(GenericModel, Generic[CommandParamsT]):
         ),
     )
 
+    def normalize(
```
this could be due to me not part of convo's about this but if I am not mistaken we are now adding `normalize` for commands. shouldnt this be just for commands that have a location prop? although there is a default body. just wondering about this.

SyntaxColoring replied:
After some refactors to implement [another to-do](https://github.com/Opentrons/opentrons/pull/12571#discussion_r1186562519), this `normalize()` method is no longer present on `BaseCommandCreate`.

jbleon95 committed:
Normalize input slot in Deck.get_slot_definition().

TamarZanzouri commented on api/src/opentrons/types.py:
```diff
@@ -223,18 +205,60 @@ class DeckSlotName(enum.Enum):
+        """
+        return _ot2_to_ot3.get(self, self)
+
+    def to_equivalent_for_robot_type(self, robot_type: RobotType) -> DeckSlotName:
```
no biggy but should we maybe call this `map_by_robot_type`? feel free to discard this just feels more clear

SyntaxColoring replied:
Leaving it as-is for now for expedience, but happy for this to change later.

TamarZanzouri commented on api/src/opentrons/types.py:
```diff
@@ -254,6 +278,27 @@ def __str__(self) -> str:
 
 
+# fmt: off
+_slot_equivalencies = [
```
another no biggy but what if we used a list of `DeckSlotMapping` type with props for ot-2 and ot-3?

TamarZanzouri replied:
or a dict of slot alias and a `DeckSlotMapping` as the value? then if there will be another change for the OT-4 ;-) the lift will be easy

SyntaxColoring replied:
@TamarZanzouri Sorry, I'm not following what you mean by this. Can you give an example?

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -49,6 +61,16 @@ class LoadLabwareParams(BaseModel):
         # TODO: Make sure v6 JSON protocols don't do that.
     )
 
+    def canonicalize(self, robot_type: RobotType) -> LoadLabwareParams:
```
should we create a base model for `BaseLabwareLocation` and move this declaration within? `MoveLabwareParams` should also inherit and implement this.

SyntaxColoring replied:
I think this is moot now that I've moved all this logic to `deck_slot_standardization.py`.

TamarZanzouri commented on api/src/opentrons/protocol_engine/protocol_engine.py:
```diff
@@ -148,6 +148,8 @@ def add_command(self, request: commands.CommandCreate) -> commands.Command:
             RunStoppedError: the run has been stopped, so no new commands
                 may be added.
         """
+        request = request.normalize(robot_type=self.state_view.config.robot_type)
```
I see why you added it to the BaseCreate

TamarZanzouri commented on api/src/opentrons/protocol_engine/protocol_engine.py:
```diff
@@ -281,13 +283,24 @@ def add_labware_offset(self, request: LabwareOffsetCreate) -> LabwareOffset:
 
         To retrieve offsets later, see `.state_view.labware`.
         """
+        normalized_request = request.copy(
```
why do we need to do it this way and not just by constructing a new object?

SyntaxColoring replied:
Both ways are error-prone, and I think constructing the object from scratch turns out to be error-prone in more dangerous ways.

My first attempt for `loadLabware` commands looked something like this:

```python
normalized = LoadLabwareCreate.construct(
    params=original.params.normalize(robot_type)
)
```

But that hides a nasty bug. When it constructs the new `LoadLabwareCreate`, it discards the original's `intent`. The new command will have the default, value `intent=None`. This was easy to miss because `intent` doesn't appear in the `LoadLabwareCreate` definition. It's implicitly inherited from `BaseCommandCreate`.

SyntaxColoring replied:
Left a comment to this effect in `deck_slot_standardization.py`.

TamarZanzouri commented:
left a few comments. nothing critical for this pr just a few thoughts

SyntaxColoring committed:
Move labware offset standardization to a mockable function.

SyntaxColoring committed:
refactor: Delete a bunch of unused fixtures in test_protocol_engine.py.

SyntaxColoring committed:
Test that ProtocolEngine.add_command() standardizes its input.

SyntaxColoring committed:
Start on some command standardization tests.

SyntaxColoring committed:
Also test ModuleLocations.

SyntaxColoring committed:
Boundless intelligence.

SyntaxColoring committed:
Test standardization of loadModule commands.

SyntaxColoring committed:
Also test off-deck locations.

SyntaxColoring committed:
Add failing tests for moveLabware standardization.

SyntaxColoring committed:
Add todo comment for Thermocycler dodging.

SyntaxColoring committed:
Fix get_edge_path_type() not accounting for OT-3 slots.

SyntaxColoring committed:
Fix fixed trash slot.

SyntaxColoring committed:
Scaffold command standardization functions.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Merging for PR #12691, and CI fixes.

SyntaxColoring committed:
Lint/typecheck fixups.

SyntaxColoring committed:
Update create_protocol_engine() test.

SyntaxColoring committed:
Work around test_geometry.py depending on legacy Deck class.

SyntaxColoring committed:
Add docstrings to test_slot_standardization.py.

SyntaxColoring committed:
Extract into free-standing functions; implement for moveLabware.

SyntaxColoring committed:
Docstrings and comments.

SyntaxColoring commented on shared-data/python/opentrons_shared_data/deck/__init__.py:
```diff
@@ -56,12 +56,22 @@ def load_schema(version: int) -> "DeckSchema":
+    # Assume that the OT-3 deck definition has the same number of slots,
+    # and in the same order, as the OT-2.
     s = slots[slot - 1]
-    assert s["id"] == str(slot)
```
It looks like we have some helpers in `hardware_testing` that had a similar assert, which is thankfully causing those tests to fail. I need to fix those spots in `hardware_testing` and double-check if there are any others.

SyntaxColoring replied:
Fixed the one in `hardware_testing`. A grep for `slot: int` suggests that that was the only remaining problematic one, but I've ticketed follow-up as https://opentrons.atlassian.net/browse/RLAB-345.

SyntaxColoring committed:
Update DeckDataProvider test.

SyntaxColoring commented on api/src/opentrons/protocol_engine/slot_standardization.py:
```diff
@@ -0,0 +1,133 @@
+
+This module does that conversion, for any Protocol Engine input that contains a reference to a
+deck slot.
+"""
```
To do: also document this general conversion behavior at the HTTP API level.

SyntaxColoring committed:
Revert "Remove internal stringification from legacy resolve_module_location()."

This reverts commit 378b8cee1b73f94d3574a44ac0503a2c0108d71c.

This wasn't necessary for this PR. This legacy Deck class is only used in PAPIv<=2.13, non-Protocol-Engine ProtocolContexts, which means it's only used with OT-2 deck definitions, so this internal stringification was "okay." It is probably bad, but it doesn't immediately break anything, so these fixes should be left out of this PR to reduce risk.

SyntaxColoring committed:
Remove vestigial TypeVar from commands.py.

SyntaxColoring committed:
Undo unnecessary formatting changes.

SyntaxColoring committed:
Fix comment typos.

SyntaxColoring committed:
get_slot_definition(): Be explicit about id stringification.

SyntaxColoring committed:
Document module calibration storing slots as an int.

SyntaxColoring committed:
Remove problematic assert from hardware_testing, and document `slot: int` in hardware_testing.

SyntaxColoring committed:
Add tests and improve documentation for DeckSlotName.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Merging for CI fixes?

SyntaxColoring committed:
Only guarantee nice errors if the caller obeys type-checking.

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_labware_view.py:
```diff
@@ -513,9 +513,7 @@ def test_get_slot_definition_raises_with_bad_slot_name(
     subject = get_labware_view(deck_definition=ot2_standard_deck_def)
 
     with pytest.raises(errors.SlotDoesNotExistError):
-        # note: normally the typechecker should catch this, but clients may
-        # not be using typechecking or our enums
-        subject.get_slot_definition(42)  # type: ignore[arg-type]
```
`get_slot_definition()` takes a `DeckSlotName`, not an `int`. This code was testing that even if you wrongly provide an `int`, it raises a nice error.

Does anyone remember why we were trying to support this? In general, I don't think internal code like Protocol Engine is responsible for behaving gracefully when it's given invalid types. Anyway, we have `ensure_deck_slot()` now in the Python Protocol API to help validate input, which I think is a more appropriate place to handle this sort of thing.

SyntaxColoring committed:
Document auto-conversion behavior at the HTTP level.

SyntaxColoring committed:
Add missing comma.

TamarZanzouri commented on api/src/opentrons/protocol_api/deck.py:
```diff
@@ -82,47 +82,59 @@ def __len__(self) -> int:
 
         return self[west_slot] if west_slot is not None else None
 
+    # todo(mm, 2023-05-08): This is undocumented in the public PAPI, but is used in some protocols
```
:fire:

SyntaxColoring committed:
Add tests for get_edge_path_type().

SyntaxColoring committed:
Add test for get_fixed_trash_slot().

TamarZanzouri commented on api/src/opentrons/protocol_engine/slot_standardization.py:
```diff
@@ -0,0 +1,133 @@
+    return original.copy(update={"params": params})
+
+
+_standardize_command_functions: Dict[
```
Nice!

TamarZanzouri commented on api/src/opentrons/protocol_engine/state/labware.py:
```diff
@@ -400,7 +400,7 @@ def get_edge_path_type(
         left_path_criteria = mount is MountType.RIGHT and well_name in left_column
         right_path_criteria = mount is MountType.LEFT and well_name in right_column
-        labware_right_side = labware_slot in [
+        labware_right_side = labware_slot.to_ot2_equivalent() in [
```
I guess this method is only used for OT-2

SyntaxColoring replied:
It's used for both. This was a shortcut to specify the slots on the right of the deck for both OT-2 and OT-3. This is confusing and I'll change it.

TamarZanzouri commented on api/tests/opentrons/protocol_engine/test_slot_standardization.py:
```diff
@@ -0,0 +1,258 @@
+            "OT-3 Standard",
+            DeckSlotLocation(slotName=DeckSlotName.SLOT_C2),
+        ),
+        # ModuleLocations and OFF_DECK_LOCATIONs should be left alone.
```
Nice catch!

TamarZanzouri reviewed (approved):
Love this! great job! The tavern test is one of our more complicated ones. pending on hardware testing of course :-)

SyntaxColoring commented:
During testing, we found what we think are latent app-side bugs that this surfaces. See RAUT-456. This shouldn't merge until those are fixed.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

hardware-testing is failing type-checking and I don't know why. Maybe it's fixed in edge.

SyntaxColoring committed:
Explicitly name right-side OT-3 slots.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Merging for CI Fixes in PR #12789.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Integrating PR #12795.

SyntaxColoring committed:
Port Tavern test to Python so we can parametrize it over OT-2/OT-3 servers.

SyntaxColoring committed:
Leave a bunch of todo comments for hard-coded slots in PE.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Merging for PR #12815.

SyntaxColoring committed:
Remove todo comment for loading Thermocycler in slot 7.

This is okay because it's outside of Protocol Engine, so Protocol Engine will normalize it when it ingests the loadModule command.

SyntaxColoring committed:
Resolve todo about hard-coded _INSTRUMENT_ATTACH_SLOT.

Per discussion with Tamar, this is going to change in the near future to be based on coordinates, not deck slots. It's also OT-3-only. So, do the simple thing for now, and don't worry about refactoring LabwareView to handle both kinds of deck.
We're

SyntaxColoring committed:
Resolve todo for fetching the deck-appropriate middle slot for TC dodging.

SyntaxColoring committed:
Update slot pairs for Thermocycler dodging. Close RLAB-339.

SyntaxColoring committed:
Remove todo in LegacyCommandMapper.

I think this is okay to leave alone because this should only be used by PAPIv<=2.13, which should never emit the new OT-3-style deck slot IDs.

SyntaxColoring committed:
Release dem notes.

SyntaxColoring committed:
Merge branch 'edge' into flex_deck_slot_names

Resolve conflicts in api/release-notes-internal.md.

SyntaxColoring committed:
Update expected slots in ReturnTip.test.tsx.

This test uses robotType: 'OT-3 Standard', so it should expect OT-3-style slots names.

SyntaxColoring committed:
Update expected slots in PickUpTip.test.tsx.

SyntaxColoring committed:
Update expected slots in CheckItem.test.tsx.

TamarZanzouri commented on api/src/opentrons/protocol_engine/slot_standardization.py:
```diff
@@ -0,0 +1,133 @@
+        return standardize(original, robot_type)
+
+
+# Command-specific standardization:
```
perfect! thank you!

TamarZanzouri commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -95,7 +95,16 @@ class SlotTransit(NamedTuple):
-]
+}
+
+_OT3_THERMOCYCLER_SLOT_TRANSITS_TO_DODGE = {
```
not for this PR but these changes are starting to make me think of moving all specific implementation details to a designated class. for example this will be a field in a class instead of a variable. same for _RIGHT_SIDE_SLOTS.

TamarZanzouri replied:
again, this is me thinking from a DB perspective. This will probably live in a db table related to each robot

TamarZanzouri reviewed (approved):
still looks great! Max tested this a lot!

SyntaxColoring committed:
Hurl my computer and all peripherals into the ocean

SyntaxColoring commented:
There's one failing app front-end test, but I think that's an unrelated thing that's fixed in #12842.
